### PR TITLE
[FSDP] summon full params staticmethod

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #75423

Make summon_full_params a static method. We still retain the old summon_full_params as a private API `_summon_full_params` and there are a couple of callsites to this within only FSDP file, but we can remove these as well.

Differential Revision: [D35444539](https://our.internmc.facebook.com/intern/diff/D35444539/)

### CODE REVIEW

File: test/distributed/fsdp/test_fsdp_summon_full_params.py [hunk 1]
```diff
@@ -308,6 +309,29 @@ def test_summon_full_params_equivalence(self, rank0_only, offload_to_cpu):
 
         self.assertEqual(fsdp_params, params_to_compare)
 
+    @skip_if_lt_x_gpu(2)
+    def test_summon_from_non_fsdp(self):
+        class FSDPContainer(nn.Module):
+            def __init__(self, fsdp_1, fsdp_2):
+                super().__init__()
+                self.fsdp_1 = fsdp_1
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_summon_full_params.py:317]: Minor: Should we include `self.fsdp_3` and in both `model_fsdp` and `model_no_fsdp`, we initialize it as `DeterministicModel(wrap_fsdp=False)`?
Knowing the implementation, this test suffices, but to test a more general non-FSDP root module, the root may have non-FSDP children.

**Review by awgu — APPROVED** [review]
Stamp to unblock 👍🏼

File: test/distributed/fsdp/test_fsdp_summon_full_params.py [hunk 2]
```diff
@@ -349,15 +373,15 @@ def test_reshard_outside_forward_backward_iteration(self, rank0_only, offload_to
             outer_full_param_size, outer_param._full_param_padded.storage().size()
         )
         self.assertEqual(0, inner_param._full_param_padded.storage().size())
-        with model.summon_full_params(rank0_only=rank0_only, writeback=not rank0_only, offload_to_cpu=offload_to_cpu):
+        with model.summon_full_params(model, rank0_only=rank0_only, writeback=not rank0_only, offload_to_cpu=offload_to_cpu):
```
[the remark this answers is not in the document]
Comment (rohan-varma) [reply test/distributed/fsdp/test_fsdp_summon_full_params.py:376]: I think its actually better practice to call the staticmethod from the class for better readability of the code?

File: test/distributed/fsdp/test_fsdp_summon_full_params.py (hunk 1 above)
Comment (rohan-varma) [reply test/distributed/fsdp/test_fsdp_summon_full_params.py:317]: I see, do you mean add a self.fsdp_3 and `self.fsdp_3 = DeterministicModel(wrap=False)`?

Comment (awgu) [reply test/distributed/fsdp/test_fsdp_summon_full_params.py:317]: Yup.

**Pull request closed without merging**